Unit tests: Move target_h/ stubs into libraries' test doubles

LDong-Arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

Fixes: #14799

This PR moves stubs headers from `UNITTESTS/target_h/` into individual libraries' own `UNITTESTS/doubles/`, similar to what we did with `UNITTESTS/stubs/`. The CMake targets that contain those headers are changed from `mbed-headers-base` to `mbed-headers-<library>`.

Note: Even though headers previously in `target_h` are technically stubs/fakes too, they are used by not only unit tests but also _regular libraries_ when compiled for unit tests, because no target-specific HAL implementation exists in this case. In order for regular library sources to pick up `target_h` headers, those headers must

* have the same names as regular headers
* appear first in include paths

This is why those headers are part of `mbed-headers-<library>` and not `mbed-stubs-<library>`. Before this refactoring, `mbed-headers-base` which contained `target_h` was the first in all unit tests' include paths.

To avoid duplicating the above explanation in every commit, it is only in the commit message of the last commit "Unit tests: Remove redundant CMake target mbed-headers-base".

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

None

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@ARMmbed/mbed-os-core 

----------------------------------------------------------------------------------------------------------------

LDong-Arm committed:
Unit tests: Move cellular target_h stubs

We move test-specific files from the top-level UNITTESTS directory
into individual components' UNITTESTS/doubles subdirectories.

LDong-Arm committed:
Unit tests: Remove unused arm_math.h

LDong-Arm committed:
Unit tests: Move cmsis target_h stubs

This creates a new mbed-headers-cmsis which is always and only used
by mbed-headers-platform, so we add the former to the laters' linked
targets.

LDong-Arm committed:
Unit tests: Remove unused nsapi_ppp.h stub

LDong-Arm committed:
Unit tests: Move hal target_h stubs

Stubs previously in UNITTESTS/target_h/ have the same names as
regular Mbed OS headers, intending to override the latter directly.
We move hal target_h stubs into hal/tests/UNITTESTS/doubles/.

Note: In Mbed OS, the standard include format requires each header to
be prefixed with its module name, for example "hal/gpio_api.h". This
requires headers to be organized in a module directory. But unit tests
stubs for hal correspond to what a real Mbed target would have
implemented (in a non-test scenario), and targets do not currently put
headers in hal/, so we similary put stub headers directly in
hal/tests/UNITTESTS/doubles/ instead of add a hal/ subdirectory there.

LDong-Arm committed:
Unit tests: Move platform target_h stubs

Stubs previously in UNITTESTS/target_h/ have the same names as
regular Mbed OS headers, intending to override the latter directly.
We move platform target_h stubs into
platform/tests/UNITTESTS/doubles/platform/.

Note: nvic_wrapper.h is normally implemented and used by Mbed targets
as needed. But as unit tests do not have a real target, we treat it
as a stub for the platform.

LDong-Arm committed:
Unit tests: Remove unused syslimits.h

LDong-Arm committed:
Unit tests: Remove mbed.h stub from target_h

The header `mbed.h` is a convenient wrapper that pre-includes some
platform headers, for use by user applications. Libraries and tests
internal to Mbed OS should not use it, and they should explicitly
include headers they need. So a stub is not needed.

LDong-Arm committed:
Unit tests: Move rtos target_h stubs

LDong-Arm committed:
Unit tests: Move drivers target_h stubs

LDong-Arm committed:
Unit tests: Move randLIB.h

The stub header randLIB.h overrides the header of the same name in
platform/randlib/ which is an external repository vendored into
the mbed-os codebase. As the repository is synchronized regularly,
it is better not to put overrides there, so we put the randLIB.h
stub into the regular platform doubles directory.

LDong-Arm committed:
Unit tests: Remove redundant CMake target mbed-headers-base

Individual libraries' `target_h` stub headers have now all been moved
from `mbed-headers-base` to `mbed-headers-<library>`.

Note: Even though headers previously in `target_h` are technically
stubs/fakes too, they are used by not only unit tests but also regular
libraries when compiled for unit tests, because no target-specific HAL
implementation exists in this case. In order for regular library
sources to pick up `target_h` headers, those headers must

* have the same names as regular headers
* appear first in include paths

This is why those headers are part of `mbed-headers-<library>` and not
`mbed-stubs-<library>`. Before this refactoring, `mbed-headers-base`
was the first in unit tests' include paths.

LDong-Arm committed:
Unit tests: Fix style in rtos Thread.h stubs

LDong-Arm committed:
Unit tests: Fix style in hal gpio_object.h stub

LDong-Arm committed:
Unit tests: Fix style in cellular myCellularContext.h stub

rajkan01 reviewed (dismissed):
Thanks for the changes @LDong-Arm , it looks good to me and needs to fix CI

LDong-Arm committed:
Unit tests: Fix pin validation failure

The test script pinvalidate.py requires the following which are
missing in the unit test stub PinName.h:
* A comment "MBED TARGET LIST"
* `CONSOLE_TX` and `CONSOLE_RX` in the `PinName` enum

This commit adds them.

LDong-Arm reviewed (commented):
@rajkan01 @rwalton-arm: Thanks for the reviews. The unit test stub `PinNames.h` didn't satisfy the latest requirements on pin names. I've added a commit to fix that, hopefully Travis will pass!

LDong-Arm committed:
Travis: Enable maximum verbosity of pinvalidate.py

Run pinvalidate.py with -vvv to print the cause of failure, making
it easier for authors of pull requests to fix pin names.

Patater commented on platform/tests/UNITTESTS/doubles/CMakeLists.txt:
```diff
@@ -4,6 +4,9 @@
+        .
+        platform
+        platform/cxxsupport
         ${mbed-os_SOURCE_DIR}/platform/include
```
In commit "Unit tests: Move platform target_h stubs", we moved many files to `platform/tests/UNITTESTS/doubles/platform`

But we moved `nvic_wrapper.h` to `platform/tests/UNITTESTS/doubles/`. Was this inconsistency intentional? Do we need the `platform/tests/UNITTESTS/doubles/platform` folder still, rather than putting everything under `platform/tests/UNITTESTS/doubles/`? We ought to know we are in the platform folder and this is redundant, but it would depend if any files were included like `#include "platform/CThunk.h` by mentioning the folder. Does nobody depend on `nvic_wrapper.h` this way and that's why we could move it out of a folder named `platform`?

LDong-Arm replied:
Yes, `#include "platform/CThunk.h` is the preferred style of includes. Even though it's not strictly followed by some code, in my latest push I just moved `nvic_wrapper.h` into `platform/tests/UNITTESTS/doubles/platform` for consistency.

rajkan01 reviewed (dismissed):
LGTM

LDong-Arm reviewed (commented):
@Patater I just moved `nvic_wrapper.h` and updated commit message to explain why hal stubs are not in a `**/hal/` subdirectory. Thanks.

Patater reviewed (dismissed):
LGTM
